internal/app/machined/pkg/controllers/block/volume_manager.go

```diff
@@ -326,60 +337,70 @@ func (ctrl *VolumeManagerController) Run(ctx context.Context, r controller.Runti
+					ShouldCloseVolume: shouldCloseVolume,
 				},
 			); err != nil {
-				volumeStatus.PreFailPhase = volumeStatus.Phase
-				volumeStatus.Phase = block.VolumePhaseFailed
-				volumeStatus.ErrorMessage = err.Error()
+				volumeStatus.TypedSpec().PreFailPhase = volumeStatus.TypedSpec().Phase
+				volumeStatus.TypedSpec().Phase = block.VolumePhaseFailed
+				volumeStatus.TypedSpec().ErrorMessage = err.Error()
 
 				if xerrors.TagIs[volumes.Retryable](err) {
 					shouldRetry = true
 				}
 			} else {
-				volumeStatus.ErrorMessage = ""
-				volumeStatus.PreFailPhase = block.VolumePhase(0)
+				volumeStatus.TypedSpec().ErrorMessage = ""
+				volumeStatus.TypedSpec().PreFailPhase = block.VolumePhase(0)
 			}
 
-			if volumeStatus.Phase != block.VolumePhaseReady {
+			if volumeStatus.TypedSpec().Phase != block.VolumePhaseReady {
 				fullyProvisionedWave = vc.TypedSpec().Provisioning.Wave - 1
 			}
 
-			if prevPhase != volumeStatus.Phase || err != nil {
+			if prevPhase != volumeStatus.TypedSpec().Phase || err != nil {
 				fields := []zap.Field{
-					zap.String("phase", fmt.Sprintf("%s -> %s", prevPhase, volumeStatus.Phase)),
+					zap.String("phase", fmt.Sprintf("%s -> %s", prevPhase, volumeStatus.TypedSpec().Phase)),
 					zap.Error(err),
 				}
 
-				if volumeStatus.Location != "" {
-					fields = append(fields, zap.String("location", volumeStatus.Location))
+				if volumeStatus.TypedSpec().Location != "" {
+					fields = append(fields, zap.String("location", volumeStatus.TypedSpec().Location))
 				}
 
-				if volumeStatus.MountLocation != "" && volumeStatus.MountLocation != volumeStatus.Location {
-					fields = append(fields, zap.String("mountLocation", volumeStatus.MountLocation))
+				if volumeStatus.TypedSpec().MountLocation != "" && volumeStatus.TypedSpec().MountLocation != volumeStatus.TypedSpec().Location {
+					fields = append(fields, zap.String("mountLocation", volumeStatus.TypedSpec().MountLocation))
 				}
 
-				if volumeStatus.ParentLocation != "" {
-					fields = append(fields, zap.String("parentLocation", volumeStatus.ParentLocation))
+				if volumeStatus.TypedSpec().ParentLocation != "" {
+					fields = append(fields, zap.String("parentLocation", volumeStatus.TypedSpec().ParentLocation))
 				}
 
-				if len(volumeStatus.EncryptionFailedSyncs) > 0 {
-					fields = append(fields, zap.Strings("encryptionFailedSyncs", volumeStatus.EncryptionFailedSyncs))
+				if len(volumeStatus.TypedSpec().EncryptionFailedSyncs) > 0 {
+					fields = append(fields, zap.Strings("encryptionFailedSyncs", volumeStatus.TypedSpec().EncryptionFailedSyncs))
 				}
 
 				volumeLogger.Info("volume status", fields...)
 			}
 
-			allClosed = allClosed && volumeStatus.Phase == block.VolumePhaseClosed
+			// when closing, ignore META volume, we want it to stay longer, so no problem if is not closed yet
+			allClosed = allClosed && (volumeStatus.TypedSpec().Phase == block.VolumePhaseClosed || vc.Metadata().ID() == constants.MetaPartitionLabel)
```

Comment:
can't we have a way to say the closing wave for meta needs to be last, so we can do proper dependency?

Reply:
this is a bit tricky:

* at shutdown, sequencer notifies this controller that all volumes should be closed, and waits for the controller to close them
* so we should close and report that all are closed
* but at the same time META can be accessed after that, and it uses volume to find it
* so this is a bit of a hack, and probably we should do this some other way